Adds Allwinner D1 support to mainline u-boot

Jamm02 opened the pull request:
With the exception of the spi driver, the series of changes adds support for Allwinner D1 to the mainline u-boot.

smaeul committed:
tools: mkimage: Add Allwinner TOC1 support

TOC1 is an container format used by Allwinner's boot0 that can hold
multiple images. It supports encryption and signatures, but that
functionality is not implemented, only the basic "non-secure" subset.

A config file is used to provide the list of data files to include. Its
path is passed as the argument to "-d". It contains sections of the
following form:

  [name]
  file = /path/to/file
  addr = 0x12345678

Specific well-known names, such as "dtb", "opensbi", and "u-boot", are
used by the bootloader to distinguish the items inside the image.

Cover-letter:
tools: mkimage: Add Allwinner TOC1 support

The SPL port for the Allwinner D1 RISC-V SoC will probably take a while
longer than porting U-Boot proper, as none of the relevant drivers are
set up for DM in SPL. In the meantime, we are using[1][2] a fork[3] of
Allwinner's boot0 loader, which they also call "spl" in their BSP. boot0
uses this TOC1 image format.

[1]: https://linux-sunxi.org/Allwinner_Nezha#U-Boot
[2]: https://fedoraproject.org/wiki/Architectures/RISC-V/Allwinner
[3]: https://github.com/smaeul/sun20i_d1_spl
END
Series-prefix: RFC
Series-to: sunxi

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
riscv: sunxi: Add device tree for Allwinner D1 nezha

Add device tree target for Allwinner D1.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
binman: Prevent entries in a section from overlapping

Currently, if the "offset" property is given for an entry, the section's
running offset is completely ignored. This causes entries to overlap if
the provided offset is less than the size of the entries earlier in the
section. Avoid the overlap by only using the provided offset when it is
greater than the running offset.

The motivation for this change is the rule used by SPL to find U-Boot on
sunxi boards: U-Boot starts 32 KiB after the start of SPL, unless SPL is
larger than 32 KiB, in which case U-Boot immediately follows SPL.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
clk: sunxi: Add support for the D1 CCU

Since the D1 CCU binding is defined, we can add support for its
gates/resets, following the pattern of the existing drivers.

Series-to: sunxi

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
pinctrl: sunxi: Add support for the D1

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
riscv: cpu: Add cache operations for T-HEAD CPUs

As Allwinner D1 integrates the 64-bit C906 core of Ali T-Head
it requires specific cache operations, this commit adds the
required cache operations and memory map.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
sunxi: Add sunxi based riscv board support

Adds basic board support for sunxi boards based on riscv.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
sunxi: Add support for Allwinner D1 Nezha

Add defconfig file for the allwinner D1 Nezha

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
drivers: Add support for sunxi SOC's with RISC-V cores

Adds common driver support to Allwinner D1.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
include: Add support for Allwinner D1 SoC

Adds header files for sun20i(Allwinner D1) for clock,
resets and mailbox in include/dt-bindings.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
riscv:dts: Add common regulators for Allwinner D1

Adds common regulators namely  gpio, codec, hdmi, etc for
sun20i(Alwinner D1).

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
riscv: dts: Add dts for Allwinner D1 Nezha

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
riscv: dts: Add Allwinner D1 SoC dtsi

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
net: sun8i-emac: Allwinner D1 Support

While R40 puts the EMAC syscon register at a different address from
other variants, the relevant portion of the register's layout is the
same. Factor out the register offset so the same code can be shared
by all variants. This matches what the Linux driver does.

This change provides two benefits beyond the simplification:
 - R40 boards now respect the RX delays from the devicetree
 - This resolves a warning on architectures where readl/writel
   expect the address to have a pointer type, not phys_addr_t.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteeenshah.02@gmail.com>

smaeul committed:
gpio: sunxi: Adds support for the Allwinner D1

Allows Allwinner D1 to use arm specific gpio driver without
implementing new drivers for arch/riscv for sunxi boards.
The implementation can be made seamless once all arm boards
drops the use of arch-specific headers.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteeenshah.02@gmail.com>

smaeul committed:
mmc: sunxi: Add support for Allwinner D1.

Allows Allwinner D1 to use arm specific mmc driver without
implementing new drivers for arch/riscv for sunxi boards.
The implementation can be made seamless once all arm boards
drops the use of arch-specific headers.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteeenshah.02@gmail.com>

smaeul committed:
usb: musb-new: Add support for Allwinner D1.

Allows Allwinner D1 to use arm specific usb driver without
implementing new drivers for arch/riscv for sunxi boards.
The implementation can be made seamless once all arm boards
drops the use of arch-specific headers.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

Icenowy commented:
Your patches still miss Signed-off-by by the original authors.

Icenowy commented:
and commit message should be consistent, among your own commits, and with existing commits.

Icenowy commented on arch/riscv/Kconfig:
```diff
@@ -28,6 +28,10 @@ config TARGET_SIPEED_MAIX
 	bool "Support Sipeed Maix Board"
 	select SYS_CACHE_SHIFT_6
 
+config TARGET_SUNXI
```
Should we try to get some consistency between arch/arm sunxi and arch/riscv one?

Jamm02 replied:
There has to be a distinguishing factor between D1 and other boards since we are not using arch-specific headers for the D1, once all the arch-specific headers are dropped for sunxi boards it can be consistent with arch-sunxi following the semantics after headers are dropped.

Icenowy commented:
Your signed-off-by should coexist with the original one, and come latter.

Jamm02 commented:
@Icenowy, I have updated the commit messages according to previous suggestions.

Jamm02 commented:
@liushuyu @Icenowy is there anything else I should add to the commits/code?

liushuyu commented:
> @liushuyu @Icenowy is there anything else I should add to the commits/code?

I am not quite sure why you pinged me as I am not a mentor of your project, not even an off-mentor.

But anyways, I will do a very basic pass of review purely on the code organization and basic code hygiene.

liushuyu commented on arch/riscv/cpu/thead/cache.c:
```diff
@@ -0,0 +1,119 @@
+	csr_set(CSR_MHCR, MHCR_IE | MHCR_RS | MHCR_BPE | MHCR_BTB | MHCR_L0BTB);
+	csr_set(CSR_MHINT, MHINT_IPLD | MHINT_IWPE);
+}
+
+void icache_disable(void)
+{
+	csr_clear(CSR_MHCR, MHCR_IE);
+}
+
+int icache_status(void)
+{
+	return csr_read(CSR_MHCR) & MHCR_IE;
+}
+
+void dcache_enable(void)
+{
+	asm volatile (".long 0x0020000b" ::: "memory"); /* dcache.iall */
+	sync_i();
+	csr_set(CSR_MHCR, MHCR_DE | MHCR_WA | MHCR_WB | MHCR_WBR);
+	csr_set(CSR_MHINT, MHINT_DPLD | MHINT_AMR_LIMIT_3);
+}
+
+void dcache_disable(void)
+{
+	asm volatile (".long 0x0010000b" ::: "memory"); /* dcache.call */
+	sync_i();
+	csr_clear(CSR_MHCR, MHCR_DE);
+}
+
+int dcache_status(void)
+{
+	return csr_read(CSR_MHCR) & MHCR_DE;
+}
+
+void enable_caches(void)
+{
+	icache_enable();
+	dcache_enable();
+}
+#endif
```
Why is this commented out? If the code is unused and you are done with the implementation, you should remove it.

liushuyu commented on drivers/clk/sunxi/clk_d1.c:
```diff
@@ -0,0 +1,96 @@
+	.num_resets = ARRAY_SIZE(d1_resets),
+};
+
+// unsigned int clock_get_pll6(void)
+// {
+// 	struct sunxi_ccm_reg *const ccm =
+// 		(struct sunxi_ccm_reg *)SUNXI_CCM_BASE;
+
+// 	uint32_t rval = readl(&ccm->pll6_cfg);
+// 	int n = ((rval & CCM_PLL6_CTRL_N_MASK) >> CCM_PLL6_CTRL_N_SHIFT) + 1;
+// 	int m = ((rval >> CCM_PLL6_CTRL_M_SHIFT) & CCM_PLL6_CTRL_M_MASK) + 1;
+// 	int p0 = ((rval >> CCM_PLL6_CTRL_P0_SHIFT) & CCM_PLL6_CTRL_P0_MASK) + 1;
+// 	/* The register defines PLL6-2X, not plain PLL6 */
+// 	return 24000000 / n / m / p0;
+// }
```
Ditto with the code comments. Please either fix the function or remove it entirely.

liushuyu commented on drivers/clk/sunxi/clk_ncatv2.h:
```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (C) 2021 Moteen Shah <moteenshah.02@gmail.com>
```
```suggestion
 * Copyright (C) 2022 Moteen Shah <moteenshah.02@gmail.com>
```

Also, did you author this file entirely, or is it based on Samuel Holland's implementation?

Jamm02 replied:
Samuel added the clock header file for sun50i_h6 because D1 utilises the same macros; however, if the additions are to be made generic, the macros must be defined in a distinct header file unique to D1; therefore, should I put Samuel or the original creator of the sun50i_h6 header in the copyright?

liushuyu replied:
> Samuel added the clock header file for sun50i_h6 because D1 utilises the same macros; however, if the additions are to be made generic, the macros must be defined in a distinct header file unique to D1; therefore, should I put Samuel or the original creator of the sun50i_h6 header in the copyright?

I think you need to include the copyright of everyone involved (as far as you understand it).

liushuyu commented on arch/riscv/cpu/thead/cache.c:
```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: GPL-2.0+
+
```
Copyright statements?

liushuyu commented on include/dt-bindings/mailbox/sun20i-d1-msgbox.h:
```diff
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+
```
Ditto with copyright statements.

liushuyu commented on drivers/usb/musb-new/sunxi.c:
```diff
@@ -174,13 +176,15 @@ static void USBC_ForceVbusValidToHigh(__iomem void *base)
 
 static void USBC_ConfigFIFO_Base(void)
 {
+	#ifndef CONFIG_TARGET_SUNXI
```
Inconsistent indentation.

Icenowy commented on Kconfig:
```diff
@@ -12,6 +12,8 @@ source "scripts/Kconfig.include"
 # Allow defaults in arch-specific code to override any given here
 source "arch/Kconfig"
 
+source "board/sunxi/Kconfig"
```
This is quite dirty and should not happen.

Jamm02 replied:
made the required changes.

Icenowy commented on drivers/pinctrl/sunxi/Kconfig:
```diff
@@ -1,6 +1,6 @@
 # SPDX-License-Identifier: GPL-2.0
 
-if ARCH_SUNXI
+if ARCH_SUNXI || TARGET_SUNXI
```
Consider use ARCH_SUNXI for both ARM and RISCV?

Jamm02 replied:
Apologies for the late reply I was occupied with my college mid-semester exams, since we are not using the arch headers there has to be a way to distinguish D1 from others that are using arch-specific headers, I found TARGET_SUNXI to be a way but for the seamless merger, it will require the arch headers.

Jamm02 replied:
also should I make a riscv/sunxi same as arm/mach-sunxi

Jamm02 replied:
Marking this as resolved for now owing to the previous conversations in the above threads.

Icenowy commented on arch/riscv/lib/cache.c:
```diff
@@ -20,7 +20,7 @@ __weak void flush_dcache_range(unsigned long start, unsigned long end)
 }
 
-void invalidate_icache_range(unsigned long start, unsigned long end)
+__weak void invalidate_icache_range(unsigned long start, unsigned long end)
```
Should this be in a splitted commit?

Jamm02 replied:
The change(adding __weak for other functions with the same name to override the definition) essentially affects the same feature,  I can still split it if advised.

Icenowy replied:
I think you'd split it.

Jamm02 replied:
done the required changes

Icenowy replied:
Your sequence of two commits are reversed, thus your first commit won't pass build.

Icenowy replied:
Your sequence is okay now, but the commit message is total mess. Please rework it to be more generic, and represent what __weak does.

Jamm02 replied:
done with the changes.

Icenowy commented on board/sunxi/Kconfig:
```diff
@@ -22,3 +41,112 @@ config SPL_IMAGE_TYPE
 	default "sunxi_egon" if SPL_IMAGE_TYPE_SUNXI_EGON
 	default "sunxi_toc0" if SPL_IMAGE_TYPE_SUNXI_TOC0
+
+if TARGET_SUNXI
```
Should we just move these things from arch/arm/mach-sunxi/Kconfig instead of guard it here with RISCV-specific option?

Jamm02 replied:
Yes I'll do the needful

Jamm02 replied:
marking this as resolved, added a new Kconfig specific to D1, which can be merged into the current Kconfig once all the arch-specific headers are dropped until then there has to be a different file or a guard to distinguish the D1(Note we are not using arch-specific headers for the D1)

Icenowy commented on Kconfig:
```diff
@@ -458,7 +458,7 @@ config BUILD_TARGET
 	default "u-boot-elf.srec" if RCAR_GEN3
 	default "u-boot.itb" if !BINMAN && SPL_LOAD_FIT && (ARCH_ROCKCHIP || \
-				ARCH_SUNXI || RISCV || ARCH_ZYNQMP)
+				RISCV || ARCH_ZYNQMP)
```
I think this breaks arm sunxi.

Jamm02 replied:
well arch sunxi uses binman internally so there wont be a case of !binman && arch_sunxi, anyways removed it.

Icenowy commented on board/sunxi/Kconfig.sun20i:
```diff
@@ -0,0 +1,148 @@
+if TARGET_SUNXI
```
Sorry but I still want you to split this.

Jamm02 replied:
I have split it into 3 commits, have a look.

Icenowy replied:
Sorry but this is not I wanted, I want you to split this from arch/arm/mach-sunxi/Kconfig, not split here.

Jamm02 replied:
changed it to use current Kconfig.

Icenowy commented on common/spl/Kconfig:
```diff
@@ -1473,7 +1473,7 @@ config SPL_OPTEE_IMAGE
 config SPL_OPENSBI
 	bool "Support RISC-V OpenSBI"
-	depends on RISCV && SPL_RISCV_MMODE && RISCV_SMODE
+	depends on RISCV && SPL_RISCV_MMODE && RISCV_SMODE && SPL_LOAD_FIT
```
Why bother add SPL changes if you do not support SPL now?

Jamm02 replied:
have removed it

Icenowy commented on configs/nezha_defconfig:
```diff
@@ -1,21 +1,67 @@
 CONFIG_RISCV=y
+CONFIG_SYS_MEMTEST_START=0x40000000
+CONFIG_SYS_MEMTEST_END=0x60000000
```
Why memtest?

Jamm02 replied:
That was added during the testing and development phase of Allwinner D1 and can be removed, but if someone wants to run their own tests it is functional.

smaeul committed:
riscv: Allow other CPU types to redefine cache function.

Declaring invalidate_icache_range as weak, allows CPU specific
definition of the invalidate_icache_range to override the
definition in lib/cache.c thus allowing other CPU types to
implement the function with its own definition.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

Icenowy reviewed (changes requested):
Your sequence of commits is a mess here too.

Icenowy commented:
Please keep the bisectibility of all these codes. (that means, if I checkout any intermediate commits of this PR, it should still pass build)

Icenowy commented on board/sunxi/Kconfig.sun20i:
```diff
@@ -108,4 +108,41 @@ config SYS_CPU
+	select SPL_OF_CONTROL if SPL_DM
+	select SPL_PINCTRL if SPL_DM
+	select SPL_SEPARATE_BSS if SPL
+	select SUPPORT_SPL
```
Are you sure you implemented SPL?

Icenowy commented:
Many commit messages need rework to be more detailed and explain technical details of the commits.

smaeul committed:
Kconfig: Refactor sunxi Kconfig

Refactors mach-sunxi/Kconfig to allow RISCV based sunxi boards
to use board/sunxi/Kconfig.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
Kconfig: Add support for USB configurations.

Adds support for VBUS_PIN, VBUS_DET, ID for USB0, USB1, USB2, USB3.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
riscv: Kconfig: Add support for Allwinner sunxi SOCs with RISCV core

Adds a new configuration for Allwinner RISC-V SOC's with board
specific configurations. Allwinner D1 cannot utilise ARCH SUNXI
because it employs arch-specific headers, and the sunxi boards
are being changed to discontinue the use of arch-specific headers.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
sunxi: Add support for SUN20I architecture

Add support for sun20i architecture which is a newer riscv64
SoC's by Allwinner.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

Icenowy commented on arch/riscv/Kconfig:
```diff
@@ -28,6 +28,44 @@ config TARGET_SIPEED_MAIX
+	select SPL_OF_CONTROL if SPL_DM
+	select SPL_PINCTRL if SPL_DM
+	select SPL_SEPARATE_BSS if SPL
+	select SUPPORT_SPL
```
Still, SUPPORT_SPL?! Are you sure?

Jamm02 replied:
yes it has been implemented

Icenowy commented on include/configs/sunxi-common.h:
```diff
@@ -33,8 +36,19 @@
 # define CONFIG_SYS_NS16550_COM5		SUNXI_R_UART_BASE
 #endif
 
+/* I2C */
```
Still a mess, I doubt this commit should be at least 3, or maybe 5, or maybe more commits.

Jamm02 replied:
splitted

Icenowy commented on arch/riscv/dts/sun20i-d1-nezha.dts:
```diff
@@ -0,0 +1,341 @@
+	status = "okay";
+
+	bluetooth {
+		compatible = "xradio,xr829-bt";
```
Device w/o legal DT binding.

Jamm02 replied:
added the legalities

Icenowy commented:
Finally you get able to write some code that I expect you to write in Day 1, however this snippet of code only values a few person hour, and tomorrow is the deadline, so I doubt whether you can finish it on time.

I suggest you to close this PR and give up.

Jamm02 commented:
> Finally you get able to write some code that I expect you to write in Day 1, however this snippet of code only values a few person hour, and tomorrow is the deadline, so I doubt whether you can finish it on time.
> 
> I suggest you to close this PR and give up.

Any additional suggestions will be really appreciated, as you believe the code is logically sound; any other suggestions can be incorporated within a day.
I have worked hard on the project for the past four months, and closing the PR is not an option for me. I am willing to work beyond the project's official timeline until you believe the PR is mergeable. I have tested all of the commits individually, and none of them breaks any other boards. Regards for your valuable time.

liushuyu commented:
> I am willing to work beyond the project's official timeline until you believe the PR is mergeable.

As a mentor of a different project in this organization, I need to clarify that it is against OSPP's rules to request extensions this late in the timeline. In this case, you might not receive any stipend/compensation for your work.

You must have the resolve to finish this project, turning things around within several days, or prepare for a disappointing result. OSPP this year will only count PR successfully merged before the deadline toward your project contribution.

This year's deadline is October 31, 2022, at 23:59:59 (UTC+8:00) [for your convenience, it should be 21:29:59 Monday, October 31, 2022, in Indian Time (UTC+5:30)].

Jamm02 commented:
> As a mentor of a different project in this organization, I need to clarify that it is against OSPP's rules to request extensions this late in the timeline. In this case, you might not receive any stipend/compensation for your work.

The reason I took on this project was not for the stipend but for the experience; my focus is to complete the project and have it merged; the compensation is secondary.

smaeul committed:
sunxi: i2c: Refactor i2c clock definitions

Refactors clock definition from arch-sunxi/i2c.h to
configs/sunxi-common.h so that both - riscv and arm
based sunxi boards can use the definition.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
include: Add support for riscv based sunxi boards

Adds support for riscv based sunxi boards (Allwinner D1)
to sunxi-common.h for RAM addresses.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
Kconfig: Add support for MACH_SUN20I

Adds SYS_TEXT_BASE configuration for Allwinner D1, we
should use the smaller address like V3S to support D1s.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
sunxi: Kconfig: Add support for MACH_SUN20I

- Adds basic support for RAM adresses and size for Allwinner D1.
- Defines the type of SOC, board and configuration for MACH_SUN20I.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

smaeul committed:
riscv: Kconfig: Add common board configuration for Allwinner D1.

Sources the common kconfig file that houses common board and
CPU configurations for MACH_SUN20I.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Moteen Shah <moteenshah.02@gmail.com>

Icenowy commented:
Sorry but I now think you didn't get what a SPL is.

Jamm02 commented:
> Sorry but I now think you didn't get what a SPL is.

Without the SPL changes introduced, it won't be able to parse the input tree.

Icenowy commented:
> > Sorry but I now think you didn't get what a SPL is.
> 
> Without the SPL changes introduced, it won't be able to parse the
> input tree.

Then where is your SPL? Is it usable?

If not, why do you set SUPPORT_SPL?

Jamm02 commented:
> > Sorry but I now think you didn't get what a SPL is. Without the SPL changes introduced, it won't be able to parse the input tree.
> Then where is your SPL? Is it usable? If not, why do you set SUPPORT_SPL?

it builds the SPL image already configured in board/sunxi/Kconfig, [this](https://github.com/u-boot/u-boot/commit/a51715d07a78890a9f0331daebf373c2100a9abd) commit allows it to use the images

Icenowy commented:
Sorry but finally this does not contain as much useful production as needed. Closing.

Jamm02 commented:
> Sorry but finally this does not contain as much useful production as needed. Closing.

Please provide a clear explanation of why it is not being merged because the goal of the project as mentioned on the OSPP site is being accomplished, I have demonstrated that the series of changes add support to Allwinner D1 and do not break any other board.

liushuyu commented:
> Please provide a clear explanation of why it is not being merged

Although I am not your mentor, I have monitored this project for a while. Let me give you my preliminary thoughts on this.

> because the goal of the project as mentioned on the OSPP site is being accomplished

First of all, the requirements clearly stated that "Basic bootloader functions, can boot a Linux kernel, and **upstreaming all relevant code changes**" however the last part was clearly not done. Your mentor gated your patches because they will need to vouch for your changes when you submit them to the U-boot upstream. When your code quality is subpar, they can't let the unfinished code slip upstream.

> I have demonstrated that the series of changes add support to Allwinner D1 and do not break any other board

I have discussed with your mentor the project status and progression. It seems to me that there are some issues with communication and mutual understanding. You seem to have trouble understanding some basic concepts you thought you had well understood, like how SPL (Secondary Program Loader) works.

It is unfortunate that it has to end like this. To reflect on this, we also have our own issues, where we should identify the communication problems early and help you manage the time and project progression better.

Speaking of which, may I ask you why you chose this project despite its advanced difficulty (given that you might have become aware that a more challenging project is more likely to fail)?

Jamm02 commented:
> > Please provide a clear explanation of why it is not being merged
> 
> Although I am not your mentor, I have monitored this project for a while. Let me give you my preliminary thoughts on this.
> 
> > because the goal of the project as mentioned on the OSPP site is being accomplished
> 
> First of all, the requirements clearly stated that "Basic bootloader functions, can boot a Linux kernel, and **upstreaming all relevant code changes**" however the last part was clearly not done. Your mentor gated your patches because they will need to vouch for your changes when you submit them to the U-boot upstream. When your code quality is subpar, they can't let the unfinished code slip upstream.
> 
> > I have demonstrated that the series of changes add support to Allwinner D1 and do not break any other board
> 
> I have discussed with your mentor the project status and progression. It seems to me that there are some issues with communication and mutual understanding. You seem to have trouble understanding some basic concepts you thought you had well understood, like how SPL (Secondary Program Loader) works.
> 
> It is unfortunate that it has to end like this. To reflect on this, we also have our own issues, where we should identify the communication problems early and help you manage the time and project progression better.
> 
> Speaking of which, may I ask you why you chose this project despite its advanced difficulty (given that you might have become aware that a more challenging project is more likely to fail)?

Thanks for the detailed review, the reason I took on the project was for interest in computer architecture in general and my previous experience in building embedded projects, so I started with this project as it complimented my skills thereby I was able to complete the "boot the Linux kernel" part of the project, but I wasn't a seasoned open source contributor (second year undergrad degree when I joined the project) so I expected some guidance from the mentors as I have seen my peers in GSOC and OSPP having weekly meets with their respective mentors regarding the project but I was wrong on my expectations. I had my previous contributions merged in red hat organizations before applying for the project under OSPP. On the topic of subpar quality of code. I updated my mentor with regular commits so she could have an overall view of the final product. I still feel the project would have been mergeable if proper communication from the start was established.
Nevertheless thank you for your time and effort.
